api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -2,38 +2,241 @@ package messagebuffer
 
 import (
 	"encoding/hex"
+	"fmt"
+	"io"
 	"sync"
+	"time"
 
+	"github.com/nknorg/nkn/v2/config"
+	"github.com/pbnjay/memory"
+
+	rbt "github.com/emirpasic/gods/trees/redblacktree"
 	"github.com/nknorg/nkn/v2/pb"
 )
 
+const maxSeqId int64 = 1000000
+
+var maxClientMsgCacheSize = float64(1 << 20) // 1M
+
+type sequence struct {
+	sync.RWMutex
+	seqId int64
+}
+
+var seq sequence
+
+func getNextSeq() int64 {
+
+	seq.Lock()
+	defer seq.Unlock()
+	seq.seqId++
+
+	if seq.seqId >= maxSeqId {
+		seq.seqId = 0
+	}
+
+	return seq.seqId
+}
+
+// msg node with create time and expire time
+type msgNode struct {
+	msg      *pb.Relay
+	createId int64
+	expireId int64
+}
+
+func (m *msgNode) GetSize() int {
+	// m.createId(int64), m.expireId(int64), m.msg.MaxHoldingSeconds(int32), m.msg.SigChainLen(int32)
+	return 8 + 8 + 4 + 4 + len(m.msg.SrcIdentifier) + len(m.msg.SrcPubkey) + len(m.msg.DestId) +
+		len(m.msg.Payload) + len(m.msg.BlockHash) + len(m.msg.LastHash)
+}
+
+type indexNode struct {
+	clientID string
+	createId int64
+	expireId int64
+}
+
 // MessageBuffer is the buffer to hold message for clients not online
 type MessageBuffer struct {
 	sync.Mutex
-	buffer map[string][]*pb.Relay
+	buffer       map[string]*rbt.Tree // map ClinetId to msgs
+	totalBufSize int                  // the buffer size in bytes
+	treeCreate   *rbt.Tree            // index tree of create time
+	treeExpire   *rbt.Tree            // index tree of expire time
+}
+
+// comparation function of red and black tree
+func compare(a, b interface{}) int {
+	aInt := a.(int64)
+	bInt := b.(int64)
+	if aInt < bInt {
+		return -1
+	} else if aInt == bInt {
+		return 0
+	} else {
+		return 1
+	}
 }
 
 // NewMessageBuffer creates a MessageBuffer
 func NewMessageBuffer() *MessageBuffer {
-	return &MessageBuffer{
-		buffer: make(map[string][]*pb.Relay),
+
+	msgBuffer := &MessageBuffer{
+		buffer:     make(map[string]*rbt.Tree), // *pb.Relay
+		treeCreate: rbt.NewWith(compare),
+		treeExpire: rbt.NewWith(compare),
 	}
+	// set max cache size by config
+	if config.Parameters.MaxClientMsgCachePercent > 0 {
+		maxClientMsgCacheSize = float64(memory.TotalMemory()) * (float64(config.Parameters.MaxClientMsgCachePercent) / 100.0)
+	}
+
+	go msgBuffer.BufferMonitor()
+
+	return msgBuffer
 }
 
 // AddMessage adds a message to message buffer
 func (messageBuffer *MessageBuffer) AddMessage(clientID []byte, msg *pb.Relay) {
+
+	if msg.MaxHoldingSeconds <= 0 { // no need buffer
+		return
+	}
+
+	now := time.Now().UnixMilli()
+	seqId := getNextSeq()
+	createId := now*maxSeqId + seqId
+	expireId := now*maxSeqId + (int64)(msg.MaxHoldingSeconds*1000) + seqId
```

Comment:
this is still ordered by create time, now expiration time

Reply:
Yes. treeMsg is ordered by createId.